src/main/java/io/supertokens/cronjobs/bulkimport/ProcessBulkImportUsers.java

```diff
@@ -0,0 +1,597 @@
+                    createUserRoles(main, appIdentifier, bulkImportProxyStorage, user);
+
+                    // We are updating the primaryUserId in the bulkImportUser entry. This will help us handle the inconsistent transaction commit.
+                    // If this update statement fails then the outer transaction will fail as well and the user will simpl be processed again. No inconsistency will happen in this
+                    // case.
+                    baseTenantStorage.updateBulkImportUserPrimaryUserId(appIdentifier, user.id,
+                            primaryLM.superTokensUserId);
+
+                    // We need to commit the transaction manually because we have overridden that in the proxy storage
+                    // If this fails, the primaryUserId will be updated in the bulkImportUser but it wouldn’t actually exist.
+                    // When processing the user again, we'll check if primaryUserId exists with the same email. In this case the user won't exist, and we'll simply re-process it.
+                    bulkImportProxyStorage.commitTransactionForBulkImportProxyStorage();
+
+                    // NOTE: We need to use the baseTenantStorage as bulkImportProxyStorage could have a different storage than the baseTenantStorage
+                    // If this fails, the primaryUserId will be updated in the bulkImportUser and it would exist in the database.
+                    // When processing the user again, we'll check if primaryUserId exists with the same email. In this case the user will exist, and we'll simply delete the entry.
+                    baseTenantStorage.startTransaction(con2 -> {
+                        baseTenantStorage.deleteBulkImportUser_Transaction(appIdentifier, con2, user.id);
+                        return null;
+                    });
+                    return null;
+                } catch (StorageTransactionLogicException e) {
+                    // We need to rollback the transaction manually because we have overridden that in the proxy storage
+                    bulkImportProxyStorage.rollbackTransactionForBulkImportProxyStorage();
+                    throw e;
+                } finally {
+                    closeAllProxyStorages();
+                }
+            });
+        } catch (StorageTransactionLogicException | InvalidBulkImportDataException e) {
+            handleProcessUserExceptions(appIdentifier, user, e, baseTenantStorage);
+        }
+    }
+
+    private void handleProcessUserExceptions(AppIdentifier appIdentifier, BulkImportUser user, Exception e,
+            BulkImportSQLStorage baseTenantStorage)
+            throws StorageQueryException {
+
+        // Java doesn't allow us to reassign local variables inside a lambda expression
+        // so we have to use an array.
+        String[] errorMessage = { e.getMessage() };
+
+        if (e instanceof StorageTransactionLogicException) {
```

Comment:
we should also add the error stack in the message? So the saved error can look like:
{
   errorMsg: "...",
   errorStack: "..."
}

Reply:
There is a utils function somewhere in the code to get the error stack from an exception as a string

Reply:
I don't think we should provide error stack. We don't expect our users to debug the issue by looking at the code. The error message should be clear enough to understand which field has the issue.

Reply:
oh yea, but users may ask on support that they are getting this error and then for us, it will be way easier to know from where. This can be skipped only if the error messages are globally unique.

Reply:
We decided to add error codes in the error message to make them globally unique.